File: lib/server/restaurant-scoring.js
/**
 * Restaurant Diagnostic Score — 9-indicator model
 *
 * Composite restaurant_score: 0–100  (Bayesian-adjusted)
 * Assessment levels (cohort-relative percentile bands):
 *   EXCELLENT  top 10%   (frac < 0.10)
 *   GOOD       10–30%    (frac < 0.30)
 *   MODERATE   30–60%    (frac < 0.60)
 *   LOW        60–80%    (frac < 0.80)
 *   RISKY      bottom 20%
 *
 * Indicator weights  (must sum to 1.00):
 *   D1  口碑星级  Rating Quality       0.20
 *   D2  评价量级  Review Volume        0.15
 *   D3  口碑情感  Review Sentiment     0.15
 *   D4  食品安全  Food Safety          0.15  ← restaurant-exclusive (DOH grade)
 *   D5  信息完整度 Profile Completeness 0.12
 *   D6  服务广度  Service Breadth      0.10
 *   D7  价值定位  Price-Value Ratio    0.08
 *   D8  运营活跃  Operational Activity  0.03
 *   D9  转化就绪  Conversion Readiness  0.02
 */

"use strict";

// ─── helpers ──────────────────────────────────────────────────────────────────

function clamp(v, lo, hi) {
  return Math.max(lo, Math.min(hi, Number(v) || 0));
}

function round1(v) {
  return Math.round(clamp(v, 0, 100) * 10) / 10;
}

// ─── D1  口碑星级  Rating Quality (0–100) ─────────────────────────────────────
// Restaurant ratings cluster 3.5–5.0; map that band to 0–100.
// Below 3.5 → 0, 5.0 → 100.
function ratingScore(rating) {
  const r = clamp(rating, 1, 5);
  return round1(Math.max(0, (r - 3.5) / 1.5) * 100);
}

// ─── D2  评价量级  Review Volume (0–100) ──────────────────────────────────────
// Restaurants plateau higher than salons (~2 000 reviews).
// log10(n+1) / log10(2000) gives ~50 at 44 reviews, ~75 at 316, ~100 at 2000.
const VOLUME_PLATEAU = 2000;
function volumeScore(reviewCount) {
  const n = Math.max(0, Math.floor(Number(reviewCount) || 0));
  return round1((Math.log10(n + 1) / Math.log10(VOLUME_PLATEAU)) * 100);
}

// ─── D3  口碑情感  Review Sentiment (0–100) ───────────────────────────────────
// Uses star-histogram if available; falls back to headline rating.
// Red-alert: ≥1 one-star → cap at 60; ≥3 one-stars → cap at 45.
function sentimentScore(reviewsDistribution, rating) {
  const dist = reviewsDistribution && typeof reviewsDistribution === "object"
    ? reviewsDistribution : {};
  const one   = Number(dist.oneStar)   || 0;
  const two   = Number(dist.twoStar)   || 0;
  const three = Number(dist.threeStar) || 0;
  const four  = Number(dist.fourStar)  || 0;
  const five  = Number(dist.fiveStar)  || 0;
  const total = one + two + three + four + five;

  let raw;
  if (total > 0) {
    const avgStar = (one + two * 2 + three * 3 + four * 4 + five * 5) / total;
    raw = ((avgStar - 1) / 4) * 100;                // 1★→0, 5★→100
  } else {
    // fallback from headline rating
    const r = clamp(rating || 0, 1, 5);
    raw = ((r - 1) / 4) * 100;
  }

  // Red-alert caps
  if (one >= 3) raw = Math.min(raw, 45);
  else if (one >= 1) raw = Math.min(raw, 60);

  return round1(clamp(raw, 0, 100));
}

// ─── D4  食品安全  Food Safety (0–100) ────────────────────────────────────────
// Restaurant-exclusive: NYC DOH inspection grade + numeric score.
// Grade baseline: A=100, B=70, C=40, Z/N/pending=50
// Penalty: each point of DOH score above 13 (A threshold) deducts 0.8 pts.
// Null grade → neutral 50.
function foodSafetyScore(latestGrade, latestScore) {
  const gradeBase = { A: 100, B: 70, C: 40, Z: 50, N: 50 };
  const base = gradeBase[String(latestGrade || "").toUpperCase()] ?? 50;

  // Inspection score: 0–13 = A territory, 14–27 = B, 28+ = C
  // Reward scores below median (11) and penalise above A threshold (13).
  const doh = Number(latestScore);
  let penalty = 0;
  if (Number.isFinite(doh) && doh > 0) {
    penalty = Math.max(0, doh - 13) * 0.8;
  }
  return round1(clamp(base - penalty, 0, 100));
}

// ─── D5  信息完整度  Profile Completeness (0–100) ─────────────────────────────
// 7 signals with unequal weights (total 100).
//   menu_url   22   (diners decide on the menu)
//   phone      18   (direct contact)
//   website    16
//   opening_hours 16
//   price      14   (helps diners plan)
//   categories 8    (discovery)
//   image_url  6    (visual trust)
function profileScore(p) {
  const has = (v) => v !== null && v !== undefined && String(v).trim().length > 0;
  return round1(
    (has(p.bestMenuUrl || p.best_menu_url)  ? 22 : 0) +
    (has(p.phone)                           ? 18 : 0) +
    (has(p.website)                         ? 16 : 0) +
    (has(p.openingHours || p.opening_hours) ? 16 : 0) +
    (has(p.price)                           ? 14 : 0) +
    (has(p.categories) && (p.categories?.length || 0) > 0 ? 8 : 0) +
    (has(p.imageUrl || p.image_url)         ?  6 : 0)
  );
}

// ─── D6  服务广度  Service Breadth (0–100) ────────────────────────────────────
// 5 service options; each adds to score.
// Weights reflect diner importance: Delivery 25, Takeout 20, Dine-in 25,
// Table service 15, Reservation 15.
function serviceScore(additionalInfo) {
  const ai = additionalInfo && typeof additionalInfo === "object" ? additionalInfo : {};

  function flag(section, key) {
    const items = Array.isArray(ai[section]) ? ai[section] : [];
    for (const item of items) {
      if (item && typeof item === "object" && key in item) return item[key] === true;
    }
    return null; // not mentioned
  }

  const delivery    = flag("Service options", "Delivery");
  const takeout     = flag("Service options", "Takeout");
  const dineIn      = flag("Service options", "Dine-in");
  const tableSvc    = flag("Dining options",  "Table service");
  const reservation = flag("Planning",        "Accepts reservations") ??
                      flag("Planning",        "Reservations recommended");

  let score = 0;
  if (delivery    !== false) score += delivery    === true ? 25 : 12;  // absent = neutral half
  if (takeout     !== false) score += takeout     === true ? 20 : 10;
  if (dineIn      !== false) score += dineIn      === true ? 25 : 12;
  if (tableSvc    !== false) score += tableSvc    === true ? 15 :  7;
  if (reservation !== false) score += reservation === true ? 15 :  7;

  return round1(clamp(score, 0, 100));
}

// ─── D7  价值定位  Price-Value Ratio (0–100) ──────────────────────────────────
// Maps price tier + star rating to a "value perception" index.
// Formula: value = ratingScore × priceMultiplier
//   $     1.20  (cheap + high rating = great value)
//   $$    1.00
//   $$$   0.80  (premium needs justification)
//   $$$$  0.65
// Raw strings like "$10–20" → treated as $, "$20–60" → $$, "$30–100" → $$$ etc.
function valueScore(price, rating) {
  const rScore = ratingScore(rating) / 100;  // 0–1

  const p = String(price || "").trim();
  let mult = 1.0;
  if      (p === "$" || /^\$1/.test(p) || p === "$1–10") mult = 1.20;
  else if (p === "$$" || /^\$1[0-9]/.test(p))            mult = 1.00;
  else if (p === "$$$" || /^\$[23][0-9]/.test(p))        mult = 0.80;
  else if (p === "$$$$" || /^\$[4-9]/.test(p) || /^\$1[0-9]{2}/.test(p)) mult = 0.65;

  return round1(clamp(rScore * mult * 100, 0, 100));
}

// ─── D8  运营活跃  Operational Activity (0–100) ───────────────────────────────
// Checks the restaurant is still operating and discoverable.
//   Not permanently closed  40
//   Not temporarily closed  30
//   Has opening hours       20
//   Has popular times data  10
function opsScore(p) {
  if (p.permanentlyClosed || p.permanently_closed) return 0;
  let score = 40;
  if (!(p.temporarilyClosed || p.temporarily_closed)) score += 30;
  const hasHours = p.openingHours || p.opening_hours;
  if (hasHours && (Array.isArray(hasHours) ? hasHours.length > 0 : true)) score += 20;
  const pt = p.popularTimesHistogram || p.popular_times_histogram;
  if (pt && typeof pt === "object" && Object.keys(pt).length > 0) score += 10;
  return round1(clamp(score, 0, 100));
}

// ─── D9  转化就绪  Conversion Readiness (0–100) ───────────────────────────────
// Three direct-action signals: phone (50), menu (30), reservation (20).
function conversionScore(p) {
  const has = (v) => v !== null && v !== undefined && String(v).trim().length > 0;
  return round1(
    (has(p.phone)                                ? 50 : 0) +
    (has(p.bestMenuUrl || p.best_menu_url)       ? 30 : 0) +
    (has(p.reserveTableUrl || p.reserve_table_url) ? 20 : 0)
  );
}

// ─── Composite restaurant_score ───────────────────────────────────────────────
// Weights:  D1 0.20 · D2 0.15 · D3 0.15 · D4 0.15 · D5 0.12
//           D6 0.10 · D7 0.08 · D8 0.03 · D9 0.02
//
// Bayesian adjustment — same pattern as salon model:
//   conf  = 1 − exp(−reviewCount / 80)   # 50% at ~55 reviews, 90% at ~184
//   prior = 0.50
//   score = (prior + (blend/100 − prior) × (0.30 + 0.70 × conf)) × 100
const WEIGHTS = {
  D1: 0.20, D2: 0.15, D3: 0.15, D4: 0.15, D5: 0.12,
  D6: 0.10, D7: 0.08, D8: 0.03, D9: 0.02,
};

function calcRestaurantScore(dims, reviewCount) {
  const blend =
    dims.D1 * WEIGHTS.D1 + dims.D2 * WEIGHTS.D2 + dims.D3 * WEIGHTS.D3 +
    dims.D4 * WEIGHTS.D4 + dims.D5 * WEIGHTS.D5 + dims.D6 * WEIGHTS.D6 +
    dims.D7 * WEIGHTS.D7 + dims.D8 * WEIGHTS.D8 + dims.D9 * WEIGHTS.D9;

  const n    = Math.max(0, Math.floor(Number(reviewCount) || 0));
  const conf = 1 - Math.exp(-n / 80);
  const prior = 0.50;
  const adjusted = prior + (blend / 100 - prior) * (0.30 + 0.70 * conf);

  return Math.round(clamp(adjusted * 100, 0, 100) * 10) / 10;
}

// ─── Assessment level (cohort-relative) ───────────────────────────────────────
// frac = rank / total, sorted desc by restaurant_score.
function assessmentLevel(fracFromTop) {
  const f = Number(fracFromTop);
  if (!Number.isFinite(f) || f < 0) return "MODERATE";
  if (f < 0.10) return "EXCELLENT";
  if (f < 0.30) return "GOOD";
  if (f < 0.60) return "MODERATE";
  if (f < 0.80) return "LOW";
  return "RISKY";
}

// ─── Main entry: score one restaurant ─────────────────────────────────────────
/**
 * @param {object} restaurant  row from info_gather_restaurants / restaurant_info_gather
 * @param {object} profile     row from info_gather_google_profiles / restaurant_google_profiles
 * @returns {object} all dimension scores + restaurant_score (ready to upsert)
 */
function scoreRestaurant(restaurant, profile) {
  const p = profile || {};
  const r = restaurant || {};

  const D1 = ratingScore(p.rating);
  const D2 = volumeScore(p.reviews_count ?? p.reviewCount);
  const D3 = sentimentScore(p.reviews_distribution ?? p.reviewsDistribution, p.rating);
  const D4 = foodSafetyScore(r.latest_grade, r.latest_score);
  const D5 = profileScore(p);
  const D6 = serviceScore(p.additional_info ?? p.additionalInfo);
  const D7 = valueScore(p.price, p.rating);
  const D8 = opsScore(p);
  const D9 = conversionScore(p);

  const dims = { D1, D2, D3, D4, D5, D6, D7, D8, D9 };
  const restaurantScore = calcRestaurantScore(dims, p.reviews_count ?? p.reviewCount ?? 0);

  return {
    restaurant_id:       p.restaurant_id ?? r.id,
    camis:               p.camis ?? r.camis,
    place_id:            p.place_id,
    rating:              p.rating,
    review_count:        p.reviews_count ?? p.reviewCount ?? 0,
    latest_grade:        r.latest_grade,
    latest_score:        r.latest_score,
    dim_rating_score:    D1,
    dim_volume_score:    D2,
    dim_sentiment_score: D3,
    dim_food_safety_score: D4,
    dim_profile_score:   D5,
    dim_service_score:   D6,
    dim_value_score:     D7,
    dim_ops_score:       D8,
    dim_conversion_score: D9,
    restaurant_score:    restaurantScore,
    match_confidence:    p.match_confidence,
    match_status:        p.match_status,
    scored_at:           new Date().toISOString(),
  };
}

module.exports = {
  ratingScore, volumeScore, sentimentScore, foodSafetyScore,
  profileScore, serviceScore, valueScore, opsScore, conversionScore,
  calcRestaurantScore, assessmentLevel, scoreRestaurant,
  WEIGHTS,
};

File: pipelines/restaurant_scoring.py
"""
Restaurant Diagnostic Score — Python implementation (mirrors restaurant-scoring.js).

9 indicators, Bayesian-adjusted composite 0–100.

Indicator weights:
  D1  口碑星级  Rating Quality        0.20
  D2  评价量级  Review Volume         0.15
  D3  口碑情感  Review Sentiment      0.15
  D4  食品安全  Food Safety           0.15  ← DOH grade (restaurant-exclusive)
  D5  信息完整度 Profile Completeness  0.12
  D6  服务广度  Service Breadth       0.10
  D7  价值定位  Price-Value Ratio     0.08
  D8  运营活跃  Operational Activity   0.03
  D9  转化就绪  Conversion Readiness   0.02
"""

import math
import os
import time
from typing import Optional
import requests

# ─── helpers ──────────────────────────────────────────────────────────────────

def clamp(v, lo=0.0, hi=100.0):
    try:
        return max(lo, min(hi, float(v)))
    except (TypeError, ValueError):
        return lo


def r1(v):
    return round(clamp(v) * 10) / 10


# ─── D1  口碑星级  Rating Quality ─────────────────────────────────────────────

def rating_score(rating) -> float:
    """Restaurant ratings cluster 3.5–5.0; map to 0–100."""
    r = clamp(rating or 0, 1, 5)
    return r1(max(0.0, (r - 3.5) / 1.5) * 100)


# ─── D2  评价量级  Review Volume ──────────────────────────────────────────────

VOLUME_PLATEAU = 2000  # restaurants plateau higher than salons (900)

def volume_score(review_count) -> float:
    n = max(0, int(review_count or 0))
    return r1(math.log10(n + 1) / math.log10(VOLUME_PLATEAU) * 100)


# ─── D3  口碑情感  Review Sentiment ───────────────────────────────────────────

def sentiment_score(reviews_distribution, rating=None) -> float:
    dist = reviews_distribution or {}
    one   = int(dist.get("oneStar",   0) or 0)
    two   = int(dist.get("twoStar",   0) or 0)
    three = int(dist.get("threeStar", 0) or 0)
    four  = int(dist.get("fourStar",  0) or 0)
    five  = int(dist.get("fiveStar",  0) or 0)
    total = one + two + three + four + five

    if total > 0:
        avg_star = (one + two*2 + three*3 + four*4 + five*5) / total
        raw = ((avg_star - 1) / 4) * 100
    else:
        r = clamp(rating or 0, 1, 5)
        raw = ((r - 1) / 4) * 100

    # Red-alert caps
    if one >= 3:
        raw = min(raw, 45)
    elif one >= 1:
        raw = min(raw, 60)

    return r1(clamp(raw))


# ─── D4  食品安全  Food Safety (DOH grade) ────────────────────────────────────

GRADE_BASE = {"A": 100, "B": 70, "C": 40, "Z": 50, "N": 50}

def food_safety_score(latest_grade, latest_score) -> float:
    base = GRADE_BASE.get(str(latest_grade or "").upper(), 50)
    try:
        doh = int(latest_score)
        penalty = max(0, doh - 13) * 0.8   # each point above A-threshold costs 0.8
    except (TypeError, ValueError):
        penalty = 0
    return r1(clamp(base - penalty))


# ─── D5  信息完整度  Profile Completeness ─────────────────────────────────────

def profile_score(p: dict) -> float:
    def has(v):
        if v is None: return False
        if isinstance(v, (list, dict)): return len(v) > 0
        return str(v).strip() != ""

    return r1(
        (22 if has(p.get("best_menu_url"))  else 0) +
        (18 if has(p.get("phone"))          else 0) +
        (16 if has(p.get("website"))        else 0) +
        (16 if has(p.get("opening_hours"))  else 0) +
        (14 if has(p.get("price"))          else 0) +
        ( 8 if has(p.get("categories"))     else 0) +
        ( 6 if has(p.get("image_url"))      else 0)
    )


# ─── D6  服务广度  Service Breadth ────────────────────────────────────────────

def _flag(additional_info, section, key):
    """Return True/False/None (None = not mentioned)."""
    items = (additional_info or {}).get(section, [])
    if not isinstance(items, list):
        return None
    for item in items:
        if isinstance(item, dict) and key in item:
            return bool(item[key])
    return None


def service_score(additional_info) -> float:
    delivery    = _flag(additional_info, "Service options", "Delivery")
    takeout     = _flag(additional_info, "Service options", "Takeout")
    dine_in     = _flag(additional_info, "Service options", "Dine-in")
    table_svc   = _flag(additional_info, "Dining options",  "Table service")
    reservation = (_flag(additional_info, "Planning", "Accepts reservations") or
                   _flag(additional_info, "Planning", "Reservations recommended"))

    score = 0
    score += 25 if delivery    is True else (12 if delivery    is None else 0)
    score += 20 if takeout     is True else (10 if takeout     is None else 0)
    score += 25 if dine_in     is True else (12 if dine_in     is None else 0)
    score += 15 if table_svc   is True else ( 7 if table_svc   is None else 0)
    score += 15 if reservation is True else ( 7 if reservation is None else 0)
    return r1(clamp(score))


# ─── D7  价值定位  Price-Value Ratio ──────────────────────────────────────────

def value_score(price, rating) -> float:
    r_score = rating_score(rating) / 100  # 0–1
    p = str(price or "").strip()
    if p in ("$",) or p.startswith("$1") or p == "$1–10":
        mult = 1.20
    elif p == "$$" or (p.startswith("$1") and "–" in p):
        mult = 1.00
    elif p == "$$$" or any(p.startswith(f"${i}") for i in range(20, 40)):
        mult = 0.80
    elif p == "$$$$" or any(p.startswith(f"${i}") for i in range(40, 1000)):
        mult = 0.65
    else:
        mult = 1.00
    return r1(clamp(r_score * mult * 100))


# ─── D8  运营活跃  Operational Activity ──────────────────────────────────────

def ops_score(p: dict) -> float:
    if p.get("permanently_closed"):
        return 0.0
    score = 40
    if not p.get("temporarily_closed"):
        score += 30
    if p.get("opening_hours"):
        score += 20
    pt = p.get("popular_times_histogram") or {}
    if isinstance(pt, dict) and pt:
        score += 10
    return r1(clamp(score))


# ─── D9  转化就绪  Conversion Readiness ─────────────────────────────────────

def conversion_score(p: dict) -> float:
    def has(v):
        return v is not None and str(v).strip() != ""
    return r1(
        (50 if has(p.get("phone"))             else 0) +
        (30 if has(p.get("best_menu_url"))     else 0) +
        (20 if has(p.get("reserve_table_url")) else 0)
    )


# ─── Composite restaurant_score ───────────────────────────────────────────────

WEIGHTS = dict(D1=0.20, D2=0.15, D3=0.15, D4=0.15, D5=0.12,
               D6=0.10, D7=0.08, D8=0.03, D9=0.02)

assert abs(sum(WEIGHTS.values()) - 1.0) < 1e-9, "Weights must sum to 1.0"


def calc_restaurant_score(dims: dict, review_count) -> float:
    blend = sum(dims[k] * WEIGHTS[k] for k in WEIGHTS)
    n = max(0, int(review_count or 0))
    conf = 1 - math.exp(-n / 80)   # 50% at ~55 reviews, 90% at ~184
    prior = 0.50
    adjusted = prior + (blend / 100 - prior) * (0.30 + 0.70 * conf)
    return round(clamp(adjusted * 100) * 10) / 10


def assessment_level(frac_from_top: float) -> str:
    """Cohort-relative percentile band."""
    f = float(frac_from_top) if frac_from_top is not None else 0.5
    if f < 0.10: return "EXCELLENT"
    if f < 0.30: return "GOOD"
    if f < 0.60: return "MODERATE"
    if f < 0.80: return "LOW"
    return "RISKY"


def score_restaurant(restaurant: dict, profile: dict) -> dict:
    """Score one restaurant. Returns a dict ready to upsert into diagnostic table."""
    p = profile or {}
    r = restaurant or {}

    D1 = rating_score(p.get("rating"))
    D2 = volume_score(p.get("reviews_count", 0))
    D3 = sentiment_score(p.get("reviews_distribution"), p.get("rating"))
    D4 = food_safety_score(r.get("latest_grade"), r.get("latest_score"))
    D5 = profile_score(p)
    D6 = service_score(p.get("additional_info"))
    D7 = value_score(p.get("price"), p.get("rating"))
    D8 = ops_score(p)
    D9 = conversion_score(p)

    dims = dict(D1=D1, D2=D2, D3=D3, D4=D4, D5=D5, D6=D6, D7=D7, D8=D8, D9=D9)
    restaurant_score = calc_restaurant_score(dims, p.get("reviews_count", 0))

    return {
        "restaurant_id":        p.get("restaurant_id") or r.get("id"),
        "camis":                p.get("camis") or r.get("camis"),
        "place_id":             p.get("place_id"),
        "rating":               p.get("rating"),
        "review_count":         p.get("reviews_count", 0),
        "latest_grade":         r.get("latest_grade"),
        "latest_score":         r.get("latest_score"),
        "dim_rating_score":     D1,
        "dim_volume_score":     D2,
        "dim_sentiment_score":  D3,
        "dim_food_safety_score": D4,
        "dim_profile_score":    D5,
        "dim_service_score":    D6,
        "dim_value_score":      D7,
        "dim_ops_score":        D8,
        "dim_conversion_score": D9,
        "restaurant_score":     restaurant_score,
        "match_confidence":     p.get("match_confidence"),
        "match_status":         p.get("match_status"),
    }


# ─── Batch runner ─────────────────────────────────────────────────────────────

def run_batch(limit=None, dry_run=False):
    """Fetch from Supabase, compute scores, print summary."""
    SUPABASE_URL = os.getenv("SUPABASE_URL", "https://supabase.360ai.link")
    KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY",
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9."
        "eyJyb2xlIjoic2VydmljZV9yb2xlIiwiaXNzIjoic3VwYWJhc2UiLCJpYXQiOjE3NjI0MDUyMDAsImV4cCI6MTkyMDE3MTYwMH0."
        "0Scyjnrqt727pMYFEP5n-MBF3OcL2SyDUhgUTSLHLCE")
    headers = {"apikey": KEY, "Authorization": f"Bearer {KEY}"}

    # fetch profiles (join camis → restaurant for DOH data)
    page, rows = 0, []
    page_size = min(limit or 500, 500)
    print("Fetching google profiles …")
    while True:
        resp = requests.get(
            f"{SUPABASE_URL}/rest/v1/info_gather_google_profiles",
            headers={**headers, "Range-Unit": "items",
                     "Range": f"{page}-{page + page_size - 1}"},
            params={"order": "created_at.asc"},
            timeout=30)
        resp.raise_for_status()
        batch = resp.json()
        if not batch: break
        rows.extend(batch)
        print(f"  {len(rows)} profiles …", end="\r")
        if len(batch) < page_size or (limit and len(rows) >= limit): break
        page += page_size
        time.sleep(0.05)
    print(f"  {len(rows)} profiles fetched.   ")

    # build camis → restaurant lookup
    print("Fetching restaurant DOH data …")
    rest_resp = requests.get(
        f"{SUPABASE_URL}/rest/v1/info_gather_restaurants",
        headers={**headers, "Range-Unit": "items", "Range": "0-9999"},
        params={"select": "id,camis,latest_grade,latest_score", "order": "camis.asc"},
        timeout=60)
    rest_resp.raise_for_status()
    rest_by_camis = {r["camis"]: r for r in rest_resp.json()}

    # score
    results = []
    for p in rows:
        r = rest_by_camis.get(p.get("camis"), {})
        results.append(score_restaurant(r, p))

    # sort and assign cohort levels
    results.sort(key=lambda x: x["restaurant_score"], reverse=True)
    n = len(results)
    for i, rec in enumerate(results):
        rec["assessment_level"] = assessment_level(i / max(n, 1))

    # summary
    scores = [r["restaurant_score"] for r in results]
    levels = {}
    for r in results:
        levels[r["assessment_level"]] = levels.get(r["assessment_level"], 0) + 1

    print(f"\n{'='*55}")
    print(f"Scored {n:,} restaurants")
    print(f"Score:  min={min(scores):.1f}  max={max(scores):.1f}  "
          f"median={sorted(scores)[n//2]:.1f}")
    print("Assessment distribution:", levels)
    print()
    print(f"{'Indicator':<28} {'Mean':>6}  {'Weight':>7}")
    print("-" * 44)
    for k, w in WEIGHTS.items():
        col = {
            "D1": "dim_rating_score", "D2": "dim_volume_score",
            "D3": "dim_sentiment_score", "D4": "dim_food_safety_score",
            "D5": "dim_profile_score", "D6": "dim_service_score",
            "D7": "dim_value_score", "D8": "dim_ops_score",
            "D9": "dim_conversion_score",
        }[k]
        name = {
            "D1": "口碑星级 Rating Quality",
            "D2": "评价量级 Review Volume",
            "D3": "口碑情感 Sentiment",
            "D4": "食品安全 Food Safety",
            "D5": "信息完整度 Profile",
            "D6": "服务广度 Service Breadth",
            "D7": "价值定位 Price-Value",
            "D8": "运营活跃 Ops Activity",
            "D9": "转化就绪 Conversion",
        }[k]
        mean_val = sum(r[col] for r in results) / max(n, 1)
        print(f"  {name:<26} {mean_val:>6.1f}   {w*100:>5.0f}%")

    if dry_run:
        print("\n[dry-run] Sample top-5:")
        for rec in results[:5]:
            print(f"  {rec['restaurant_score']:>5.1f}  {rec['assessment_level']:<10}  "
                  f"grade={rec['latest_grade']}  rating={rec['rating']}  "
                  f"n={rec['review_count']}")
    return results


if __name__ == "__main__":
    import argparse
    p = argparse.ArgumentParser()
    p.add_argument("--limit", type=int, default=500)
    p.add_argument("--dry-run", action="store_true")
    args = p.parse_args()
    run_batch(limit=args.limit, dry_run=args.dry_run)
